internal/storage/prefixdb.go

```diff
@@ -0,0 +1,632 @@
+//
+// Do not modify the contents of the arguments while the returned iterator is in use.
+//
+// It implements the [DB] interface for type PrefixDB.
+func (pDB *PrefixDB) Iterator(start, end []byte) (Iterator, error) {
+	itStart, itEnd, err := prefixedIteratorBounds(pDB.prefix, start, end)
+	if err != nil {
+		return nil, fmt.Errorf("prefixed DB namespace reverse iterator: %w", err)
+	}
+
+	pDB.mu.Lock()
+	defer pDB.mu.Unlock()
+
+	it, err := pDB.db.Iterator(itStart, itEnd)
+	if err != nil {
+		return nil, fmt.Errorf("prefixed DB namespace iterator: %w", err)
+	}
+
+	return newPrefixDBIterator(pDB.prefix, start, end, it)
+}
+
+// ReverseIterator returns an iterator over a domain of keys, in descending
+// order. The caller must call Close when done. End is exclusive, and start must
+// be less than end. A nil end iterates from the last key (inclusive), and a nil
+// start iterates to the first key (inclusive). Empty keys are not valid.
+// No writes may happen within a domain of keys while an iterator exists over it.
+//
+// Do not modify the contents of the arguments while the returned iterator is in use.
+//
+// It implements the [DB] interface for type PrefixDB.
+func (pDB *PrefixDB) ReverseIterator(start, end []byte) (Iterator, error) {
+	itStart, itEnd, err := prefixedIteratorBounds(pDB.prefix, start, end)
+	if err != nil {
+		return nil, fmt.Errorf("prefixed DB namespace reverse iterator: %w", err)
+	}
+
+	pDB.mu.Lock()
+	defer pDB.mu.Unlock()
+
+	it, err := pDB.db.ReverseIterator(itStart, itEnd)
+	if err != nil {
+		return nil, fmt.Errorf("prefixed DB namespace reverse iterator: %w", err)
+	}
+
+	return newPrefixDBIterator(pDB.prefix, start, end, it)
+}
+
+// NewBatch creates a batch for atomic database updates.
+// The caller is responsible for calling Batch.Close() once done.
+//
+// It implements the [DB] interface for type PrefixDB.
+func (pDB *PrefixDB) NewBatch() Batch {
+	pDB.mu.Lock()
+	defer pDB.mu.Unlock()
+
+	return newPrefixDBBatch(pDB.prefix, pDB.db.NewBatch())
+}
+
+// Compact compacts the specified range of keys in the database.
+//
+// It implements the [DB] interface for type PrefixDB.
+func (pDB *PrefixDB) Compact(start, end []byte) error {
+	if err := pDB.db.Compact(start, end); err != nil {
+		return fmt.Errorf("prefixed DB namespace compaction: %w", err)
+	}
+	return nil
+}
+
+// Close closes the database connection.
+// It is not safe to close a DB until all outstanding iterators are closed
+// or to call Close concurrently with any other DB method. It is not valid
+// to call any of a DB's methods after the DB has been closed.
+//
+// It implements the [DB] interface for type PrefixDB.
+func (pDB *PrefixDB) Close() error {
+	pDB.mu.Lock()
+	defer pDB.mu.Unlock()
+
+	return pDB.db.Close()
+}
+
+// Print prints all the key/value pairs in the database for debugging purposes.
+//
+// It implements the [DB] interface for type PrefixDB.
+func (pDB *PrefixDB) Print() error {
```

Comment:
Shouldn't we change this to accept an `io.Writer` for greater flexibility of where to output the prints? This would be a breaking change though.

Reply:
Not at the moment but note it down as potential improvmenet.